lib/bolt/plugin.rb

```diff
@@ -288,16 +313,55 @@ def resolve_single_reference(reference)
 
       validate_proc.call(reference)
 
-      begin
+      result = begin
         # Evaluate the plugin and then recursively evaluate any plugin returned by it.
         hook.call(reference)
       rescue StandardError => e
         loc = "resolve_reference in #{plugin_name}"
         raise PluginError::ExecutionError.new(e.message, plugin_name, loc)
       end
+
+      if cache?(reference)
+        cache = read_cache
+        cache.merge!({ id => { 'result' => result,
+                               'mtime' => Time.now,
+                               'ttl' => ttl } })
+        File.write(@config.project.cache_file, cache.to_json)
+      end
+      result
     end
     private :resolve_single_reference
 
+    def read_cache
+      if File.zero?(@config.project.cache_file)
+        {}
+      else
+        # This will fail if the file is empty
+        JSON.parse(File.read(@config.project.cache_file))
+      end
+    end
+
+    def cache?(reference)
+      if reference.key?('_cache')
+        r = reference['_cache']
+        unless r.is_a?(Hash)
+          raise Bolt::ValidationError,
+                "_cache must be a Hash, received #{r.class}: #{r.inspect}"
+        end
+
+        unless r.key?('ttl')
+          raise Bolt::ValidationError, "_cache must set 'ttl' key"
+        end
+
+        unless r['ttl'].is_a?(Integer) && r['ttl'] >= 0
+          raise Bolt::ValidationError,
+                "_cache.ttl must be a positive Integer, received #{r.class}: #{r.inspect}"
+        end
+      end
+      reference.dig('_cache', 'ttl') || @config.plugin_cache&.fetch('ttl', nil)
```

Comment:
```suggestion
      reference.dig('_cache', 'ttl') || @config.plugin_cache['ttl']
```

If `plugin-cache` has a default in config you can get rid of the safe operator.

Reply:
It shouldn't have a default though, we need to know if the user set it or not. I guess we could set plugin-cache to `ttl => nil` but I don't see how that's better?